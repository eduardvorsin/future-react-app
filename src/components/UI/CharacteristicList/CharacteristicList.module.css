.characteristic-list {
	padding: 0;
	margin: 0;
	list-style: none;
}

.characteristic-list__item {
	display: flex;
	justify-content: space-between;
	align-items: flex-end;
}

.characteristic-list__item:not(:last-child) {
	margin-bottom: 10px;
}

.characteristic-list__property {
	position: relative;
	color: var(--text-color-primary);
}

.characteristic-list__property::before {
	border-bottom: 1px dotted var(--brand-color-100);
	bottom: 0;
	content: '';
	display: block;
	left: 0;
	position: absolute;
	width: calc(100% - 10px);
}

.characteristic-list__value {
	color: var(--brand-color-200);
}

.characteristic-list__value,
.characteristic-list__property {
	flex: 0 1 50%;
}