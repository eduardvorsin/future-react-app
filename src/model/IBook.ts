import { AccessInfo, SaleInfo, VolumeInfo } from '../API/bookTypes';

export type IBook = VolumeInfo & AccessInfo & SaleInfo;

export type IBookPartial =
  { id: string } &
  Pick<VolumeInfo, 'authors' | 'maturityRating' | 'title' | 'imageLinks'> & Pick<SaleInfo, 'retailPrice'>;
