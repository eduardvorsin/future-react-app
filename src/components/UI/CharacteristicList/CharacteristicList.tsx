import React, { FC, useMemo } from 'react';
import classes from './CharacteristicList.module.css';
import { createCharacteristicsData } from '../../../helpers/helpers';
import { IBook } from '../../../model/IBook';

type CharacteristicListProps = {
  className?: string,
  bookData: IBook,
};

const CharacteristicList: FC<CharacteristicListProps> = ({
  className,
  bookData,
}) => {
  const characteristicListClasses = [
    classes['characteristic-list'],
    className,
  ].join(' ');

  const characteristics = useMemo(() => createCharacteristicsData(bookData), [bookData]);

  return (
    <ul
      className={characteristicListClasses}
    >
      {characteristics.map((characteristic) => (
        <li
          key={characteristic.name}
          className={classes['characteristic-list__item']}
        >
          <span
            className={classes['characteristic-list__property']}
          >
            {characteristic.name}
          </span>
          <span
            className={classes['characteristic-list__value']}
          >
            {characteristic.value}
          </span>
        </li>
      ))}
    </ul>
  );
};

export default CharacteristicList;
