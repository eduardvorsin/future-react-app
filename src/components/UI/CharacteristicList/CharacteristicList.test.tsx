import React from 'react';
import { render, screen } from '@testing-library/react';
import CharacteristicList from './CharacteristicList';
import { IBook } from '../../../model/IBook';

const mockData = {
  title: 'title1',
  authors: ['author1', 'author2'],
  language: 'en',
  maturityRating: 'NOT_MATURE',
  pageCount: 120,
  publishedDate: '10.12.2013',
  publisher: 'publisher',
  webReaderLink: '',
  buyLink: '',
  isEbook: true,
} as IBook;

describe('CharacteristicList tests', () => {
  it('is rendered correctly', () => {
    render(
      <CharacteristicList
        bookData={mockData}
      />,
    );

    expect(screen.getByRole<HTMLUListElement>('list')).toBeInTheDocument();
    expect(screen.getAllByRole<HTMLUListElement>('listitem')).toHaveLength(7);
    expect(screen.getByText<HTMLSpanElement>(/author1/i)).toBeInTheDocument();
  });

  it('if an empty array is passed to bookData, it should render ul without li', () => {
    render(
      <CharacteristicList
        // @ts-expect-error - checking if the array is empty
        bookData={[]}
      />,
    );

    expect(screen.getByRole<HTMLUListElement>('list')).toBeInTheDocument();
    expect(screen.queryByRole<HTMLLIElement>('listitem')).not.toBeInTheDocument();
    expect(screen.queryByText<HTMLSpanElement>('author1')).not.toBeInTheDocument();
  });

  it('is a basic snapshot', () => {
    render(
      <CharacteristicList
        bookData={mockData}
      />,
    );

    expect(screen.getByRole<HTMLUListElement>('list')).toMatchSnapshot();
  });

  it('is a snapshot with an empty list', () => {
    render(
      <CharacteristicList
        // @ts-expect-error - checking if the array is empty
        bookData={[]}
      />,
    );

    expect(screen.getByRole<HTMLUListElement>('list')).toMatchSnapshot();
  });
});
