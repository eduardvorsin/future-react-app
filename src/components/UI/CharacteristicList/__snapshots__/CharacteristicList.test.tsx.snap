// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacteristicList tests is a basic snapshot 1`] = `
<ul
  class="characteristic-list "
>
  <li
    class="characteristic-list__item"
  >
    <span
      class="characteristic-list__property"
    >
      Authors
    </span>
    <span
      class="characteristic-list__value"
    >
      author1, author2
    </span>
  </li>
  <li
    class="characteristic-list__item"
  >
    <span
      class="characteristic-list__property"
    >
      Language
    </span>
    <span
      class="characteristic-list__value"
    >
      English
    </span>
  </li>
  <li
    class="characteristic-list__item"
  >
    <span
      class="characteristic-list__property"
    >
      Content 18+
    </span>
    <span
      class="characteristic-list__value"
    >
      Yes
    </span>
  </li>
  <li
    class="characteristic-list__item"
  >
    <span
      class="characteristic-list__property"
    >
      Page count
    </span>
    <span
      class="characteristic-list__value"
    >
      120
    </span>
  </li>
  <li
    class="characteristic-list__item"
  >
    <span
      class="characteristic-list__property"
    >
      Published date
    </span>
    <span
      class="characteristic-list__value"
    >
      10.12.2013
    </span>
  </li>
  <li
    class="characteristic-list__item"
  >
    <span
      class="characteristic-list__property"
    >
      Publisher
    </span>
    <span
      class="characteristic-list__value"
    >
      publisher
    </span>
  </li>
  <li
    class="characteristic-list__item"
  >
    <span
      class="characteristic-list__property"
    >
      Book Type
    </span>
    <span
      class="characteristic-list__value"
    >
      E-book
    </span>
  </li>
</ul>
`;

exports[`CharacteristicList tests is a snapshot with an empty list 1`] = `
<ul
  class="characteristic-list "
/>
`;
